data/source/MongoDb.php

```diff
@@ -212,91 +233,98 @@ public static function enabled($feature = null) {
-			'replicaSet' => false,
-			'schema'     => null,
-			'gridPrefix' => 'fs',
-			'w'          => 1,
-			'wTimeoutMS' => 10000,
-			'readPreference' => null,
-			'autoConnect' => false,
-			'dsn' => null
+			'host'               => static::DEFAULT_HOST . ':' . static::DEFAULT_PORT,
+			'login'              => null,
+			'password'           => null,
+			'database'           => null,
+			'dsn'                => null,
+			'timeout'            => 100,
+			'uriOptions'            => [
+				'w'                        => 'majority',
+				'wTimeoutMS'               => 10000,
+				'journal'                  => true,
+				'readConcernLevel'         => 'local',
+				'readPreference'           => 'primary',
+				'readPreferenceTags'       => [],
+			],
+			'driverOptions' => [
+				'allow_invalid_hostname' => false,
+				'weak_cert_validation'   => false
+			]
 		];
-		parent::__construct($config + $defaults);
+		$config = Set::merge($defaults, $config);
+		if (!isset($config['uriOptions']['connectTimeoutMS']) && $config['timeout']) {
+			$config['uriOptions']['connectTimeoutMS'] = $config['timeout'];
+		}
+		parent::__construct($config);
 	}
 
 	/**
 	 * Initializer. Adds operator handlers which will later allow to correctly cast any
-	 * values. Constructs a DSN from configuration.
+	 * values. Constructs a DSN from configuration, if not given.
 	 *
 	 * @see lithium\data\source\MongoDb::$_operators
 	 * @see lithium\data\source\MongoDb::_operators()
 	 * @return void
 	 */
 	protected function _init() {
-		$hosts = [];
-		foreach ((array) $this->_config['host'] as $host) {
-			$host = HostString::parse($host) + [
-				'host' => static::DEFAULT_HOST,
-				'port' => static::DEFAULT_PORT
-			];
-			$hosts[] = "{$host['host']}:{$host['port']}";
-		}
-		if ($this->_config['login']) {
-			$this->_config['dsn'] = sprintf(
-				'mongodb://%s:%s@%s/%s',
-				$this->_config['login'],
-				$this->_config['password'],
-				implode(',', $hosts),
-				$this->_config['database']
-			);
-		} else {
-			$this->_config['dsn'] = sprintf(
-				'mongodb://%s',
-				implode(',', $hosts)
-			);
-		}
 		parent::_init();
 
+		switch (true) {
+
+			case (!empty($this->_config['dsn'])):
+				if (!$this->_config['database']) {
+					throw new ConfigException('No database configured.');
+				}
+				break;
+
+			case (!$this->_config['host']):
+				throw new ConfigException('No host configured.');
+
+			default:
+				$hosts = [];
+				foreach ((array) $this->_config['host'] as $host) {
+					$host = HostString::parse($host) + [
+						'host' => static::DEFAULT_HOST,
+						'port' => static::DEFAULT_PORT
+					];
+					$hosts[] = "{$host['host']}:{$host['port']}";
+				}
+				$this->_config['dsn'] = (!$this->_config['login'])
+					? sprintf('mongodb://%s', implode(',', $hosts))
+					: sprintf(
+							'mongodb://%s:%s@%s/%s',
+							$this->_config['login'],
+							$this->_config['password'],
+							implode(',', $hosts),
+							$this->_config['database']);
```

Comment:
We don't use the ternary operator over multiple lines, instead use the good 'ol `if () {} else {}`.